File: chapter 1/typescript/BinarySearch.ts
class BinarySearch {
  binarySearch(array: number[], value: number): number | null  {
    let lowerPos: number = 0;
    let higherPos: number = array.length - 1;

    while (lowerPos <= higherPos) {
      let midPos: number = Math.floor((lowerPos + higherPos) / 2);
      let guessValue: number = array[midPos];
      if (guessValue === value) {
        return midPos;
      }
      if (guessValue > value) {
        higherPos = midPos - 1;
      } else {
        lowerPos = midPos + 1;
      }
    }
    return null;
  }
}

const array: number[] = [1, 3, 5, 7, 9];
const binarySearch = new BinarySearch();

const itemIndexBinarySearch: number | null = binarySearch.binarySearch(
  array,
  9
);

if (itemIndexBinarySearch) {
  console.log(
    `Binary search: Position: ${itemIndexBinarySearch} - Value: ${array[itemIndexBinarySearch]}`
  );
}

File: chapter 1/java/BinarySearch.java
public class BinarySearch {
    private static Integer binarySearch(int[] array, int value) {
        int lowerPos = 0;
        int higherPos = array.length - 1;
        while (lowerPos <= higherPos) {
            int midPos = (lowerPos + higherPos) / 2;
            int guessValue = array[midPos];
            if (guessValue == value) {
                return midPos;
            }
            if (guessValue > value) {
                higherPos = midPos - 1;
            } else {
                lowerPos = lowerPos + 1;
            }
        }
        return null;
    }
    
    public static void main(String[] args) {
        int[] array = { 1, 3, 5, 7, 9 };
        int findedPos = binarySearch(array, 9);
        System.out.print("Position: " + findedPos + " Value: " + array[findedPos]);
    }
}
